#!/bin/bash

#########################################################
#
# Platform: NCI Gadi HPC
# Description: Runs gatk4_genotypegvcfs.sh in parallel using inputs
# created from gatk4_genotypegvcfs_make_input.sh
# These scripts perform joint-calling, producing genotypes for multiple
# samples. Output includes an unrecalibrated/filtered VCF file per interval
# which are later gathered into a single VCF
# Usage: Adjust <project> and compute resource requests following the guide
# below, then qsub gatk4_genotypegvcfs_run_parallel.pbs
# Job resource requirements:
#   ~32GB mem per task required
#   Hugemem nodes are most SU efficient, but there is a restriction
#   to max nodes that can be requested
#   1CPU per task on hugemembw allows ~36Gb mem per CPU
#     - not yet benchmarked on this node, hardware is slower than hugemem nodes
#   1CPU per task on hugemem allows ~31Gb mem per CPU
#     - walltime 00:54:06 - 02:04:35 for 34 samples, ~640 intervals 
#     - fastest to run this in chunks my splitting up inputs file into 5
#     - even "chunks", and changing the INPUTS= variable in this PBS script
#       submitting this job 5 times (my apologies for not setting it up this
#       way, it is on the to do list)
# Author: Tracy Chew
# tracy.chew@sydney.edu.au
# Date last modified: 17/08/2020
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance
# <or e.g. bioinformatics assistance of <PERSON>> of Sydney Informatics
# Hub and resources and services from the National Computational
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
#
#########################################################

#PBS -P <project>
#PBS -N genotypegvcfs
#PBS -l walltime=06:00:00,ncpus=140,mem=5100GB,wd
#PBS -q hugemembw
#PBS -W umask=022
#PBS -l storage=scratch/<project>
#PBS -o ./Logs/gatk4_genotypegvcfs.o
#PBS -e ./Logs/gatk4_genotypegvcfs.e

module load openmpi/4.0.2
module load nci-parallel/1.0.0

set -e

# NCPUs = CPUs per task
NCPUS=2

INPUTS=./Inputs/gatk4_genotypegvcfs.inputs

#########################################################
# Do not edit below this line (unless running on a node that
# does not have 48 CPU, in which case edit the value of 'CPN'
#########################################################

# SCRIPT
SCRIPT=./gatk4_genotypegvcfs.sh

CPN=48 #CPUs per node 
M=$(( CPN / NCPUS )) #tasks per node

echo "$(date) : GATK 4 GenotypeGVCFs using interval databases created with GenomicsDBImport."

sed "s|^|${SCRIPT} |" ${INPUTS} > ${PBS_JOBFS}/input-file

mpirun --np $((M * PBS_NCPUS / 48)) \
        --map-by node:PE=${NCPUS} \
        nci-parallel \
        --verbose \
        --input-file ${PBS_JOBFS}/input-file

