#!/bin/bash

#########################################################
#
# Platform: NCI Gadi HPC
# Description: Runs gatk4_genomicsdbimport.sh in parallel using inputs
# created from gatk4_genomicsdbimport_make_input.sh
# These scripts consolidate GVCFs for multiple samples at 3200 
# genomic intervals. The result is 3200 GenomicsDBImport "databases"
# required for GenotypeGVCFs (next step for joint-calling)
# Usage: Adjust <project> and compute resource requests following the guide
# below, then qsub gatk4_genomicsdbimport_run_parallel.pbs
# Job resource requirements:
# Memory increases with cohort size. 
#   NCPUS=2 (default) will allocate ~62.5GB per task on hugemem node
#   - 35GB memory per task (of 3200 tasks) required for 6 samples
#   - 48GB memory per task (of 3200 tasks) required for 27 samples 
#   - 64GB memory per task (of 3200 tasks) required for 190 samples 
#         Note: for large cohorts (e.g. n=190), you will likely need to run
#         gatk4_genomicsdbimport_missing steps for some tasks with 375Gb memory per task
#         See gatk4_genomicsdbimport_missing_run_parallel.pbs for more info
# Author: Tracy Chew
# tracy.chew@sydney.edu.au
# Date last modified: 17/08/2020
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance
# <or e.g. bioinformatics assistance of <PERSON>> of Sydney Informatics
# Hub and resources and services from the National Computational
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
#
#########################################################

#PBS -P <project>
#PBS -N genomicsdbimport
#PBS -l walltime=03:00:00,ncpus=48,mem=1500GB,wd
#PBS -q normal
#PBS -W umask=022
#PBS -l storage=scratch/<project>
#PBS -o ./Logs/gatk4_genomicsdbimport.o
#PBS -e ./Logs/gatk4_genomicsdbimport.e

module load openmpi/4.0.2
module load nci-parallel/1.0.0

set -e

# NCPUs = CPUs per task
NCPUS=2

#########################################################
# Do not edit below this line 
#########################################################

# SCRIPT
SCRIPT=./gatk4_genomicsdbimport.sh
INPUTS=./Inputs/gatk4_genomicsdbimport.inputs

echo "$(date) : GATK 4 Consolidate VCFs across samples with GenomicsDBImport."

if [[ $PBS_QUEUE =~ bw-exec ]]; then CPN=28; else CPN=48; fi
M=$(( CPN / NCPUS )) #tasks per node

sed "s|^|${SCRIPT} |" ${INPUTS} > ${PBS_JOBFS}/input-file

mpirun --np $((M * PBS_NCPUS / CPN)) \
        --map-by node:PE=${NCPUS} \
        nci-parallel \
        --verbose \
        --input-file ${PBS_JOBFS}/input-file
